File: oauth2/src/jvmTest/java/com/okta/oauth2/kmp/jvm/SessionTokenFlowTest.java
/*
 * Copyright 2022-Present Okta, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.okta.oauth2.kmp.jvm;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import com.okta.authfoundation.client.TokenInfo;
import java.util.Collections;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import org.junit.Test;

/**
 * Java-language tests for {@link SessionTokenFlow} CompletableFuture wrapper. Validates that the
 * API is ergonomic from actual Java code.
 */
public class SessionTokenFlowTest {

  @Test
  public void start_ReturnsCompletableFutureWithTokenInfo()
      throws ExecutionException, InterruptedException, TimeoutException {
    try (SessionTokenFlow flow = TestFlowFactory.createSuccessSessionTokenFlow()) {
      CompletableFuture<TokenInfo> future =
          flow.start("example-session-token", "com.example.app:/callback");
      assertNotNull("Future should not be null", future);

      TokenInfo tokenInfo = future.get(5, TimeUnit.SECONDS);
      assertNotNull("TokenInfo should not be null", tokenInfo);
      assertEquals("Bearer", tokenInfo.getTokenType());
      assertEquals("test-access-token", tokenInfo.getAccessToken());
    }
  }

  @Test
  public void start_WithAllParams_ReturnsTokenInfo()
      throws ExecutionException, InterruptedException, TimeoutException {
    try (SessionTokenFlow flow = TestFlowFactory.createSuccessSessionTokenFlow()) {
      CompletableFuture<TokenInfo> future =
          flow.start(
              "example-session-token",
              "com.example.app:/callback",
              Collections.emptyMap(),
              "openid profile email");
      TokenInfo tokenInfo = future.get(5, TimeUnit.SECONDS);
      assertNotNull("TokenInfo should not be null", tokenInfo);
    }
  }

  @Test
  public void start_WithError_CompletesExceptionally()
      throws TimeoutException, InterruptedException {
    try (SessionTokenFlow flow = TestFlowFactory.createFailingSessionTokenFlow()) {
      CompletableFuture<TokenInfo> future = flow.start("bad-token", "com.example.app:/callback");
      try {
        future.get(5, TimeUnit.SECONDS);
        fail("Should have thrown ExecutionException");
      } catch (ExecutionException e) {
        assertNotNull("Cause should not be null", e.getCause());
        assertTrue(
            "Should contain error message",
            e.getCause().getMessage().contains("OIDC Endpoints not available."));
      }
    }
  }

  @Test
  public void constructor_WithOAuth2Client_CreatesFlow() {
    com.okta.authfoundation.client.kmp.OAuth2Client kmpClient = TestOAuth2Client.create();
    SessionTokenFlow flow = new SessionTokenFlow(kmpClient);
    assertNotNull(flow);
    flow.close();
  }

  @Test
  public void close_IsIdempotent() {
    SessionTokenFlow flow = TestFlowFactory.createSuccessSessionTokenFlow();
    flow.close();
    flow.close(); // Should not throw
  }
}

File: oauth2/src/commonTest/kotlin/com/okta/oauth2/kmp/SessionTokenFlowTest.kt
/*
 * Copyright 2022-Present Okta, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.okta.oauth2.kmp

import com.okta.authfoundation.client.TokenInfo
import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class SessionTokenFlowTest {
    @Test
    fun start_WhenSuccess_ReturnsTokenInfo() =
        runTest {
            val fakeToken = FakeSessionTokenInfo()
            val mockFlow =
                object : SessionTokenFlow {
                    override suspend fun start(
                        sessionToken: String,
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                        scope: String,
                    ): Result<TokenInfo> = Result.success(fakeToken)
                }

            val result = mockFlow.start("example-session-token", "com.example.app:/callback")

            assertTrue(result.isSuccess)
            val tokenInfo = result.getOrNull()
            assertNotNull(tokenInfo)
            assertEquals("Bearer", tokenInfo.tokenType)
            assertEquals("test-access-token", tokenInfo.accessToken)
        }

    @Test
    fun start_WhenEndpointUnavailable_ReturnsFailure() =
        runTest {
            val mockFlow =
                object : SessionTokenFlow {
                    override suspend fun start(
                        sessionToken: String,
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                        scope: String,
                    ): Result<TokenInfo> = Result.failure(IllegalStateException("OIDC Endpoints not available."))
                }

            val result = mockFlow.start("example-session-token", "com.example.app:/callback")

            assertTrue(result.isFailure)
            assertTrue(result.exceptionOrNull()?.message?.contains("OIDC Endpoints not available.") == true)
        }

    @Test
    fun start_WhenNoLocationHeader_ReturnsFailure() =
        runTest {
            val mockFlow =
                object : SessionTokenFlow {
                    override suspend fun start(
                        sessionToken: String,
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                        scope: String,
                    ): Result<TokenInfo> = Result.failure(IllegalStateException("No location header in response."))
                }

            val result = mockFlow.start("example-session-token", "com.example.app:/callback")

            assertTrue(result.isFailure)
            assertTrue(result.exceptionOrNull()?.message?.contains("No location header") == true)
        }

    @Test
    fun start_WithExtraParameters_ReturnsTokenInfo() =
        runTest {
            var capturedExtraParams: Map<String, String>? = null
            val mockFlow =
                object : SessionTokenFlow {
                    override suspend fun start(
                        sessionToken: String,
                        redirectUrl: String,
                        extraRequestParameters: Map<String, String>,
                        scope: String,
                    ): Result<TokenInfo> {
                        capturedExtraParams = extraRequestParameters
                        return Result.success(FakeSessionTokenInfo())
                    }
                }

            mockFlow.start(
                sessionToken = "example-session-token",
                redirectUrl = "com.example.app:/callback",
                extraRequestParameters = mapOf("extraOne" to "bar"),
                scope = "openid profile email"
            )

            assertNotNull(capturedExtraParams)
            assertEquals("bar", capturedExtraParams["extraOne"])
        }
}

private class FakeSessionTokenInfo : TokenInfo {
    override val id: String = "test-id"
    override val clientId: String = "test-client"
    override val issuerUrl: String = "https://example.okta.com"
    override val tokenType: String = "Bearer"
    override val expiresIn: Int = 3600
    override val accessToken: String = "test-access-token"
    override val scope: String? = "openid profile"
    override val refreshToken: String? = "test-refresh-token"
    override val idToken: String? = null
    override val deviceSecret: String? = null
    override val issuedTokenType: String? = null
}
